core/modules/parser/SelectParser.cc

```diff
@@ -74,13 +107,74 @@ namespace parser {
 ////////////////////////////////////////////////////////////////////////
 class AntlrParser {
 public:
-    AntlrParser(std::string const& q)
-        : statement(q),
-          stream(q, std::stringstream::in | std::stringstream::out),
-          lexer(stream),
-          parser(lexer)
-        {}
-    void run() {
+    virtual ~AntlrParser() {}
+    virtual void setup() = 0;
+    virtual void run() = 0;
+    virtual query::SelectStmt::Ptr getStatement() = 0;
+
+protected:
+    enum State {
+        INIT, SETUP_DONE, RUN_DONE
+    };
+    std::string stateString(State s) {
+        if (s == INIT) {
+            return "INIT";
+        } else if (s == SETUP_DONE) {
+            return "SETUP_DONE";
+        } else if (s == RUN_DONE) {
+            return "RUN_DONE";
+        }
+        return "??";
+    }
+
+    void changeState(State to) {
+        switch (_state) {
+            case INIT:
+                if (SETUP_DONE != to) {
+                    throw ParseException("Parse error(INTERNAL):invalid state transition from INIT to "
+                            + stateString(to));
+                }
+                _state = SETUP_DONE;
+                break;
+
+            case SETUP_DONE:
+                if (RUN_DONE != to) {
+                    throw ParseException("Parse error(INTERNAL):invalid state transition from SETUP_DONE to "
+                            + stateString(to));
+                }
+                _state = RUN_DONE;
+                break;
+
+            case RUN_DONE:
+                // there are no valid transitions from RUN_DONE
+                throw ParseException("Parse error(INTERNAL):invalid state transition from RUN_DONE to "
+                        + stateString(to));
+                break;
```

Comment:
When an exception is thrown then the normal flow of the code gets interrupted at a point where the "throw" is called and a control is passed upstream to the closest "try" block. IN over words, that "break;" would never get a chance to be executed.

Reply:
of course. my point was that in this case it neither hurts nor helps. I do, however, see that in come cases it can cause a static analysis tool to complain. So I'll take it out.